### PR TITLE
Add Python 3.4, 3.5, and 3.6 support 

### PR DESCRIPTION
**Pull request by MatthieuDartiailh**
This work is huge, I am not sure anybody can review it in full, but anybody willing is welcome. I did my best to clean things up but this is a rework of stuff I have been working on for several months. I will do a review myself and ping when the one week delay before merge will start.

This is based on #225

### CODE REVIEW

File: enaml/core/compiler_helpers.py [hunk 1]
```diff
@@ -604,7 +605,7 @@ def make_unpack_map(node):
         instantiation.
 
     """
-    return dict(zip(node.names, node.iternodes()))
+    return dict(list(zip(node.names, node.iternodes())))
```
Comment (serprex) [inline enaml/core/compiler_helpers.py]: `list` doesn't seem necessary here

File: enaml/qt/docking/dock_manager.py [hunk 2]
```diff
@@ -169,7 +169,7 @@ def save_layout(self):
 
         """
         items = [self._dock_area] + self.floating_frames()
-        return DockLayout(*map(LayoutSaver(), items))
+        return DockLayout(*list(map(LayoutSaver(), items)))
```
Comment (serprex) [inline enaml/qt/docking/dock_manager.py]: `list` isn't necessary here

File: enaml/qt/docking/q_dock_container.py [hunk 3]
```diff
@@ -92,7 +93,7 @@ def _closestDockBar(container):
 
     tie_breakers = (3, 0, 2, 1)
     borders = (QDockBar.North, QDockBar.East, QDockBar.South, QDockBar.West)
-    values = zip(edge_ranks, edge_weights, tie_breakers, borders)
+    values = list(zip(edge_ranks, edge_weights, tie_breakers, borders))
```
Comment (serprex) [inline enaml/qt/docking/q_dock_container.py]: `list` isn't necessary here. Alternatively use the in-place sort method over sorted

File: enaml/qt/docking/xbms.py [hunk 4]
```diff
@@ -41,21 +42,21 @@ def __init__(self, width, height, data):
 
         """
         assert len(data) == (width * height)
-        bytes = []
-        for row in xrange(height):
+        bytes_list = []
```
Comment (serprex) [inline enaml/qt/docking/xbms.py]: `bytearray` seems the more suitable type here

File: enaml/workbench/ui/workbench_window.enaml [hunk 5]
```diff
@@ -32,4 +32,4 @@ enamldef WorkbenchWindow(MainWindow):
         Include:
             objects << window_model.menus
     Include:
-        objects << filter(None, [window_model.workspace.content])
+        objects << list(filter(None, [window_model.workspace.content]))
```
Comment (serprex) [inline enaml/workbench/ui/workbench_window.enaml]: Commenting here to have you double check this line. I don't feel `list` should be necessary but I don't know the context this is running in, or if this line was computer modified. If it was it left filter's func being None, or is this a manual change & this library handles None being passed to filter?

**Comment by serprex** [comment]
I made a cursory glance. Did not review the C++ code. If you run into any bugs in byteplay please create an issue, I'll see if I can help & add a testcase

**Comment by MatthieuDartiailh** [comment]
Thanks for the review @serprex. I will look into it. As far as byteplay is concerned I think I fixed all the bugs I found and submitted them back to your repo, but I will keep you posted if I find anything new.

File: enaml/workbench/ui/workbench_window.enaml (hunk 5 above)
Comment (MatthieuDartiailh) [reply enaml/workbench/ui/workbench_window.enaml]: objects needs to be a list (hence the casting for python 3) but an inline if else should work too

File: enaml/qt/docking/xbms.py (hunk 4 above)
Comment (MatthieuDartiailh) [reply enaml/qt/docking/xbms.py]: I think that Bytes does not accept bytearray (we may consider changing this in atom, but it is outside the scope of this PR).

File: enaml/qt/docking/q_dock_container.py (hunk 3 above)
Comment (MatthieuDartiailh) [reply enaml/qt/docking/q_dock_container.py]: Indeed.

File: enaml/core/compiler_helpers.py (hunk 1 above)
Comment (MatthieuDartiailh) [reply enaml/core/compiler_helpers.py]: Indeed

File: enaml/qt/docking/dock_manager.py (hunk 2 above)
Comment (MatthieuDartiailh) [reply enaml/qt/docking/dock_manager.py]: Indeed

File: enaml/core/expression_engine.py [hunk 6]
```diff
@@ -253,7 +253,7 @@ def copy(self):
         """
         new = ExpressionEngine()
         handlers = sortedmap()
-        for key, value in self._handlers.items():
+        for key, value in list(self._handlers.items()):
```
Comment (frmdstryr) [inline enaml/core/expression_engine.py]: Does list need to be there?

File: enaml/core/template.py [hunk 7]
```diff
@@ -194,7 +194,7 @@ def get_specialization(self, args):
             # any specialization which is ambiguous. The lowest score
             # wins and a tie will raise an exception.
             score = 0
-            items = zip(argspec, spec.paramspec)
+            items = list(zip(argspec, spec.paramspec))
```
Comment (frmdstryr) [inline enaml/core/template.py]: List needed here?

File: enaml/qt/qt_toolkit_object.py [hunk 8]
```diff
@@ -91,6 +91,7 @@ def destroy(self):
         widget = self.widget
         if widget is not None:
             widget.setParent(None)
+            self.widget.deleteLater()
```
Comment (frmdstryr) [inline enaml/qt/qt_toolkit_object.py]: Can just be `widget.deleteLater()` to save a lookup

File: setup.py [hunk 9]
```diff
@@ -64,16 +80,67 @@
...
     author='The Nucleic Development Team',
     author_email='sccolbert@gmail.com',
     url='https://github.com/nucleic/enaml',
     description='Declarative DSL for building rich user interfaces in Python',
     long_description=open('README.rst').read(),
-    requires=['atom', 'PyQt', 'ply', 'kiwisolver'],
-    install_requires=['distribute', 'atom >= 0.3.8', 'kiwisolver >= 0.1.2', 'ply >= 3.4'],
+    requires=['future', 'atom', 'PyQt', 'ply', 'kiwisolver', 'qtpy'],
+    install_requires=['setuptools', 'future', 'atom',
+                      'kiwisolver', 'ply >= 3.4', 'qtpy'],
```
Comment (frmdstryr) [inline setup.py]: qtpy should be in the Qt5 PR

**Comment by MatthieuDartiailh** [comment]
Thanks for the feedback @frmdstryr. I addressed all your comments.

**Comment by MatthieuDartiailh** [comment]
I forgot to mention it and failed to keep it apparent in the history but part of this work was done by @peterazmanov and I will mention him in the release notes.

File: setup.py [hunk 10]
```diff
@@ -64,16 +80,67 @@
...
+    version=__version__,
     author='The Nucleic Development Team',
     author_email='sccolbert@gmail.com',
     url='https://github.com/nucleic/enaml',
     description='Declarative DSL for building rich user interfaces in Python',
     long_description=open('README.rst').read(),
-    requires=['atom', 'PyQt', 'ply', 'kiwisolver'],
-    install_requires=['distribute', 'atom >= 0.3.8', 'kiwisolver >= 0.1.2', 'ply >= 3.4'],
+    requires=['future', 'atom', 'PyQt', 'ply', 'kiwisolver'],
+    install_requires=['setuptools', 'future', 'atom',
```
Comment (frmdstryr) [inline setup.py]: This requires atom>=0.4 and kiwisolver>=0.1.3. These should be released to pip as currently they must be built using the git repos. Building fails w/o these. 

Also I had to install 'wheel' for 'bdist_wheel' to work (python 3.5.2 on linux x86-64). Not sure if that should be in the requirements.

Comment (MatthieuDartiailh) [reply setup.py]: As those versions are not yet officialy released I would prefer to wait on them, but this is true it needs to be updated before enaml next release.

Installing wheel to build wheel distribution is normal and does not need to be added here.

**Comment by MatthieuDartiailh** [comment]
It seems that some random failures of Travis are not simply Travis related. I will investigate.

**Comment by MatthieuDartiailh** [comment]
I addressed all comments I believe I will merge two weeks after merging #225 save if somebody as more comments in the meantime.

**Comment by MatthieuDartiailh** [comment]
To avoid introducing more backward compatibility issues than necessary I renamed the parsing package to parser such that 
```python
from enaml.core.parser import parse
```
will keep working.

I will merge in about two weeks. In the meantime feel free to comment !

**Comment by MatthieuDartiailh** [comment]
Merging since I will be on holidays next week.

**Pull request merged**